docs: add OpenTracing docs (#828)
elastic/apm-agent-nodejs: docs/opentracing.asciidoc

[[opentracing]]

ifdef::env-github[]
NOTE: For the best reading experience,
please view this documentation at https://www.elastic.co/guide/en/apm/agent/nodejs/current/opentracing.html[elastic.co]
endif::[]

== OpenTracing support

> The Elastic APM OpenTracing bridge allows to create Elastic APM transactions and spans,

Reviewer: ```suggestion
The Elastic APM OpenTracing bridge allows creating Elastic APM transactions and spans,
```

using the https://opentracing-javascript.surge.sh/[OpenTracing API].
In other words,
it translates the calls to the OpenTracing API to Elastic APM and thus allows for reusing existing instrumentation.

For more information about OpenTracing see the https://opentracing.io/[OpenTracing website].

[float]
[[ot-prerequisites]]
=== Prerequisites

OpenTracing support for the Elastic APM Node.js Agent is provided via a separate module called https://www.npmjs.com/package/elastic-apm-node-opentracing[`elastic-apm-node-opentracing`].

This module requires that the Elastic APM Node.js Agent is installed separately.
To ensure that both dependencies added to the application,
install them like so:

[source,bash]
----
npm install elastic-apm-node elastic-apm-node-opentracing --save
----

[float]
[[ot-terminologies]]
=== OpenTracing vs Elastic APM terminologies

Elastic APM differentiates between {apm-overview-ref-v}/transactions.html[transactions] and {apm-overview-ref-v}/transaction-spans.html[spans].
A transaction can in the context of OpenTracing be thought of as a special kind of span.

> While OpenTracing natively only have the concept of spans,

Reviewer 2: "only has"

the Elastic APM OpenTracing bridge will automatically create either Elastic transactions or Elastic spans behind the scenes depending on a set of rules:

1. If `agent.currentTransaction` is `null`,

>    a new Elastic transaction will be created when calling `tracer.startSpan()`

Author: Funny, I was thought not to put periods at the end of sentences when show in a list. But I'll take your word for it 😅

2. If `agent.currentTransaction` holds an existing transaction,
   but that transaction is ended,
   a new Elastic transaction will be created when calling `tracer.startSpan()`
3. In all other cases,
   a new Elastic span will be created when calling `tracer.startSpan()`

[float]
[[ot-initialization]]
=== Initialization

It's important that the agent is started before you require *any* other modules in your Node.js application - i.e. before `express`, `http`, etc.

This means that you should probably require and start the agent in your application's main file (usually `index.js`, `server.js` or `app.js`).

> Here's a simple example where we first start the agent it self and the initialize the OpenTracing bridge:

Reviewer 2: "itself"
"and then"

[source,js]
----
// Add this to the VERY top of the first file loaded in your app
const agent = require('elastic-apm-node').start({
  // Override service name from package.json
  // Allowed characters: a-z, A-Z, 0-9, -, _, and space
  serviceName: '',

  // Use if APM Server requires a token
  secretToken: '',

  // Set custom APM Server URL (default: http://localhost:8200)
  serverUrl: '',
})

const Tracer = require('elastic-apm-node-opentracing')

// Pass the Elastic APM agent as an argument to the OpenTracing tracer
const tracer = new Tracer(agent)

const span = tracer.startSpan('my-first-span')
// ... do some work ...
span.finish()
----

[float]
[[ot-api]]
=== API

[source,js]
----
tracer = new Tracer(agent)
----

The `elastic-apm-node-opentracing` module exposes a Tracer class which is OpenTracing compatible.

When instantiating the Tracer object,
an instance of the Elastic APM Node.js Agent must be provided as its only argument.

For details about the `tracer` API,

> see the https://opentracing-javascript.surge.sh/[`opentracing-javescript` API docs].

Reviewer: ```suggestion
see the https://opentracing-javascript.surge.sh/[`opentracing-javascript` API docs].
```

[float]
[[ot-elastic-apm-tags]]
=== Elastic APM specific tags

> Elastic APM defines some tags which are have special meaning and which will not be stored as regular tags.

Reviewer 2: remove "are"

> Instead they will be used to set certain metadata on the transaction or span.

Reviewer: ```suggestion
Instead, they will be used to set certain metadata on the transaction or span.
```
